### PR TITLE
Handling network errors with reachability api in rx example

### PR DESCRIPTION
**Pull request by vodovozovge**
simple handling network errors in RxExample -> AutoLoading and RxExample -> WikipediaImageSearch.
Using Reachability (https://github.com/ashleymills/Reachability.swift) with reactive wrapper
Reference: https://github.com/ReactiveX/RxSwift/issues/187

### CODE REVIEW

**Comment by vodovozovge** [comment]
some updates - implemented ReachabilityService and now used in AutoLoading and Wikipedia image search examples

I have several concerns:
1. Not sure, that extension of `ObservableConvertibleType` with `retryOnBecomesReachable` is in the right place (in ReachabilityService file now). If not, where should it be?
2. I use different placeholders for loading (empty) and for retrying (the one with "loading" label) `WikipediaSearchCell`. For me if connection is ok, images are loaded quite fast, so this "loading" label doesn't make sense. But when connection is dropped with showing "loading" placeholder we can tell user, that something happens and he can wait. Is it ok?

**Comment by kzaher** [comment]
Hi @vodovozovge ,

I think that `retryOnBecomesReachable` is in ok place.

I think it would be best to have Image enum
```
enum Image {
     case Content(UIImage)
     case OfflinePlaceholder
}
```
and return `OfflinePlaceholder` when there is error fetching image.

We can create convenience observer on UIImage to bind `Image` (maybe `rxea_image`). In it we can use some centered [placeholder image similar to](http://www.outsystems.com/forgeCore/_image.aspx/A28atdGCIn2i_ZW11S_0KWgJm3iqQCfV_w==/offline-apps) in case OfflinePlaceholder or AspectFill in case `Content` is being bound.

Using some image seems more performant and closer to real world case then generating that image :)

Let me know what you think.

**Comment by vodovozovge** [comment]
hey @kzaher 
I agree that it's not good idea to generate images all the way.
I used this code generated image because didn't want to add image to assets somehow. It's not a real issue for sure, so I will search for some open source image which represent lack of connectivity (seems your example is from some commercial system) and add it to assets.

I need some time to become familiar with binding (I found some info here https://github.com/ReactiveX/RxSwift/tree/master/Documentation#bindings , but not so detailed, so I missed it. Will check how it's done in examples)

Another issue is that we already have `Image` type alias

```
#if os(iOS)
    typealias Image = UIImage
#elseif os(OSX)
    typealias Image = NSImage
#endif
```

and actually not every image is downloadable. So for not downloadable images `OfflinePlaceholder` doesn't make sense. Probably we can introduce `DownloadableImage` enum, which will be used only for such case of fallible downloads?

Btw, what `rxea` prefix means? Rx example application?

**Comment by kzaher** [comment]
Hi @vodovozovge ,

I think `DownloadableImage` is a fine name for enum.

Take a look at `rx_image` extension for `UIImage`. You can easily create similar `rxex_image` extension that even uses `rx_image` internally or does something similar after converting `DownloadableImage` to `UIImage`.

For picture, I think we can just use image of (🔴Offline) text in large font (emojis should be in vector form) or something similar so we don't have copyright issues. Maybe there are some nicer emojis.

I thought `rxex` or something similar for prefix to identify rx example app.

Take your time to study the existing extensions. No rush :)

File: RxExample/RxExample/Services/UIImageView+DownloadableImage.swift [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    func rxex_downloadableImageAnimated(transitionType:String?) -> AnyObserver<DownloadableImage> {
+
+        return AnyObserver { [weak self] event in
+            MainScheduler.ensureExecutingOnScheduler()
+
+            switch event{
+            case .Next(let value):
+                for subview in self?.subviews ?? [] {
```
Comment (vodovozovge) [inline RxExample/RxExample/Services/UIImageView+DownloadableImage.swift]: don't like this part, but should be ok for demo

**Comment by vodovozovge** [comment]
Hi @kzaher 
I updated PR with some improvements: 
1. added `DownloadableImage` enum
2. extended `ImageService` to support downloading image with retry
3. extended `UIImageView` with `rxex_downloadableImage` and `rxex_downloadableImageAnimated`. I use `UILabel` with "⚠️" symbol for image placeholder and add it to image view when necessary. 
4. WikipediaImageSearch now uses `DownloadableImage` for handling images in cells

**Comment by kzaher** [comment]
Cool :), I think this is much better.

There are some details I would like to fine tune, but I'm fine merging this :)

**Comment by vodovozovge** [comment]
@kzaher can you share your thoughts when you have time? I want to make it better :)

**Comment by kzaher** [comment]
khju

**Comment by kzaher** [comment]
(Cleaning keyboard :) Nothing drastic :) Just some details :) I'll show you the commit ...

**Comment by vodovozovge** [comment]
thanks)

**Pull request merged**